src/algorithms/digi/PulseCombiner.cc

```diff
@@ -0,0 +1,153 @@
+        for(auto pulse : newPulse) {
+          sum_pulse.addToAmplitude(pulse);
+        }
+      }
+      debug("CellID {} has {} pulses, combined into {} clusters", cellID, pulses.size(), clusters.size());
+    }
+
+  }
+
+} // PulseCombiner:process
+
+std::vector<std::vector<edm4hep::TimeSeries>> PulseCombiner::clusterPulses(const std::vector<edm4hep::TimeSeries> pulses) const {
+
+  //Clone the pulses aray of pointers so they aren't const
+  std::vector<edm4hep::TimeSeries> ordered_pulses;
+  for(auto pulse : pulses) {
+    ordered_pulses.push_back(pulse);
+  }
+
+  // Sort pulses by time, greaty simplifying the combination process
+  std::sort(ordered_pulses.begin(), ordered_pulses.end(), [](edm4hep::TimeSeries a, edm4hep::TimeSeries b) {
+    return a.getTime() < b.getTime();
+  });
+
+  // Create vector of pulses
+  std::vector<std::vector<edm4hep::TimeSeries>> cluster_pulses;
+  float clusterEndTime = 0;
+  bool isNewPulse = true;
+  // Create clusters of pulse indices which overlap with at least the minimum separation
+  for (auto pulse: ordered_pulses) {
+    float pulseStartTime = pulse.getTime();
+    float pulseEndTime = pulse.getTime() + pulse.getInterval()*pulse.getAmplitude().size();
+    if(!isNewPulse) {
+      if (pulseStartTime < clusterEndTime + m_minimum_separation) {
+        cluster_pulses.back().push_back(pulse);
+        clusterEndTime = std::max(clusterEndTime, pulseEndTime);
+      } else {
+        isNewPulse = true;
+      }
+    }
+    if(isNewPulse) {
+      cluster_pulses.push_back({pulse});
+      clusterEndTime = pulseEndTime;
+      isNewPulse = false;
+    }
+  }
+
+  return cluster_pulses;
+
+} // PulseCombiner::clusterPulses
+
+std::vector<float> PulseCombiner::sumPulses(const std::vector<edm4hep::TimeSeries> pulses) const {
+
+  // Find maximum time of pulses in cluster
+  float maxTime = 0;
+  for (auto pulse: pulses) {
+    maxTime = std::max(maxTime, pulse.getTime() + pulse.getInterval()*pulse.getAmplitude().size());
+  }
+
+  //Calculate maxTime in interval bins
+  int maxStep = (maxTime - pulses[0].getTime())/pulses[0].getInterval();
+
+  std::vector<float> newPulse(maxStep, 0.0);
+
+  for(auto pulse: pulses) {
+    //Calculate start and end of pulse in interval bins
+    int startStep = (pulse.getTime() - pulses[0].getTime())/pulse.getInterval();
+    int pulseSize = pulse.getAmplitude().size();
+    int endStep = startStep + pulseSize;
+    for(int i = 0; i < newPulse.size(); i++) {
+      // Add pulse values to new pulse
+      float contribution = 0;
+      if(i >= startStep && i < endStep){
+        contribution = pulse.getAmplitude()[i - startStep];
+      } else { // Interpolate first and last two values to extrapolate over 0
+        if(i < startStep) {
+          contribution = pulse.getAmplitude()[0] + (pulse.getAmplitude()[1] - pulse.getAmplitude()[0])*(i - startStep);
+          if(std::signbit(contribution)!=std::signbit(pulse.getAmplitude()[0])) continue;
```

Comment:
This looks a bit dangerous. If the numerical derivative comes in at a wrong sign, this will be never triggered and waveform will be filled with junk. What is the purpose of this interpolation?

Reply:
If the derivative is wrong, this should mean the pulse is moving away from 0 so either the length of the original pulse should be longer to capture the length, or noise has already been included. 

If it's never triggered the values should be 0 from that particular pulse rather than junk until the last time step from all of the pulses being added.

It could be removed but obviously brings the simulation a little closer to reality, there won't be any sudden drops to 0 from the minimum contribution included from the pulse generator. I will add a configuration to turn it off for the moment anyway.

Reply:
Sounds like something that should be addressed in the waveform generator. It should have more padding for the pulse tails to be faithfully represented.

Reply:
Ok, I have removed the interpolation functionality

Reply:
Great, thanks!